crates/solvers/src/infra/dex/balancer/mod.rs

```diff
@@ -0,0 +1,156 @@
+use {
+    crate::{
+        boundary,
+        domain::{auction, dex, eth, order},
+    },
+    ethereum_types::U256,
+    std::sync::atomic::{self, AtomicU64},
+    tracing::Instrument,
+};
+
+mod dto;
+
+/// Bindings to the Balancer Smart Order Router (SOR) API.
+pub struct Sor {
+    client: reqwest::Client,
+    endpoint: reqwest::Url,
+    vault: boundary::balancer::Vault,
+    settlement: eth::ContractAddress,
+}
+
+pub struct Config {
+    /// The URL for the Balancer SOR API.
+    pub endpoint: reqwest::Url,
+
+    /// The address of the Balancer Vault contract.
+    pub vault: Option<eth::ContractAddress>,
+
+    /// The address of the Settlement contract.
+    pub settlement: Option<eth::ContractAddress>,
+}
+
+impl Sor {
+    async fn quote(&self, query: &dto::Query) -> Result<dto::Quote, Error> {
+        let request = serde_json::to_string(&query)?;
+        tracing::trace!(endpoint = %self.endpoint, %request, "quoting");
+        let response = self
+            .client
+            .post(self.endpoint.clone())
+            .header("content-type", "application/json")
+            .body(request)
+            .send()
+            .await?
+            .error_for_status()?
+            .text()
+            .await?;
+        tracing::trace!(%response, "quoted");
+        let quote = serde_json::from_str(&response)?;
+
+        Ok(quote)
+    }
+
+    pub async fn swap(
+        &self,
+        order: &dex::Order,
+        slippage: dex::Slippage,
+        gas_price: auction::GasPrice,
+    ) -> Result<dex::Swap, Error> {
+        let query = dto::Query::from_domain(order, gas_price);
+        let quote = {
+            // Set up a tracing span to make debugging of API requests easier.
+            // Historically, debugging API requests to external DEXs was a bit
+            // of a headache.
+            static ID: AtomicU64 = AtomicU64::new(0);
+            let id = ID.fetch_add(1, atomic::Ordering::Relaxed);
+            self.quote(&query)
+                .instrument(tracing::trace_span!("quote", id = %id))
+                .await?
+        };
+
+        if quote.is_empty() {
+            return Err(Error::NotFound);
+        }
+
+        let (input, output) = match order.side {
+            order::Side::Buy => (quote.return_amount, quote.swap_amount),
+            order::Side::Sell => (quote.swap_amount, quote.return_amount),
+        };
+
+        let (max_input, min_output) = match order.side {
+            order::Side::Buy => (slippage.add(input), output),
+            order::Side::Sell => (input, slippage.sub(output)),
+        };
```

Comment:
Let me try to walk you through my thought process while reading this code, because I might be misunderstanding something.

It seems like this particular match statement should be part of the domain logic, because the way that you apply the slippage to the input/output amounts is a business rule and it's the same for all DEXs, it's not specific to just Balancer. (Is that correct?) The first thing that comes to mind is `order.apply_slippage(slippage)` but these `input` and `output` variables show that this is not the right idea. Maybe it's a step in the right direction? Maybe there should be a `dex::Quote` object with `apply_slippage`? WDYT?

Reply:
This is a good point, and I _kind_ of tried to do it. The idea is that in the `domain` you have:
- `dex::slippage::Slippage` - the general slippage parameters
- `dex::Order` - simplified orders for DEX and DEX aggregators
- `dex::slippage::Tolerance` - the actual slippage tolerance needed for an order.

The computation of all of these things are done in the `domain` - because, as you pointed out, it is business logic.

The issue is, as you mentioned that applying a slippage tolerance to a Balancer swap. Here, we add slippage to one of the swap amounts **just for computing the swap calldata** - but the actual slippage that should be used, and how it gets determined for a particular order is kept the domain. The reason I thought that this fits here in infra code, is that it is adapter logic to make the the Balancer SOR API quack like a DEX/DEXag API quacks. For example 1Inch, 0x, ParaSwap all have a "slippage" parameter that gets added to the API query and used for computing the calldata - we emulate this here for Balancer SOR as the calldata gets computed locally instead of by the API. Does this make sense?

> Maybe there should be a dex::Quote object with apply_slippage? WDYT?

I actually *almost* did this, and introduced a:

```
struct dex::Execution {
    input: eth::Asset,
    output: eth::Asset,
}
```

kind of how I broke up the `dex::Swap` and added a sub-struct `dex::Call` for use with the `boundary::balancer::Vault` abstraction. By adding a `dex::Execution`, I could have moved this match statement into a `slippage::Tolerance::apply(dex::Execution, order::Side)`. I chose against it because encoding slippage into calldata _felt_ infrastructure-y enough to keep it as is (especially considering that all other DEX/DEXag abstractions would have it this way as well). In fact, if the Balancer contracts were written a different way, it could have just been an additional parameter in the calldata for a `slippage_tolerance_bps` such that the limit amounts get computed on-chain.

Reply:
Looking at your follow-up PR and the explanation you gave here, everything seems correct to me.